### DISCUSSION TITLE
Modeling efficiency measures and CCS with `add-on` technology

### DISCUSSION THREAD

**Discussion by SongminYu** [iiasa/message_ix#847] [Q&A] [answered]
Hi MESSAGE team and everyone, I am trying to add efficiency measures and CCS as `add-on technologies` by following this [tutorial](https://github.com/iiasa/message_ix/blob/v3.8.0/tutorial/westeros/westeros_addon_technologies.ipynb). 

**First, to model efficiency measures**, taking the "clinker-making" process in cement production as an example:
- the parent technology is called `clinker_making`
- for each Mt of clinker produced, the energy `input` includes
  - coal: a MWa
  - electricity: b MWa

There are multiple add-on technologies that can improve the efficiency of `clinker_making`, i.e., reducing the input of coal and electricity per Mt of cement production. However, this is different from the [tutorial](https://github.com/iiasa/message_ix/blob/v3.8.0/tutorial/westeros/westeros_addon_technologies.ipynb), where the output of the parent technology is used to produce another side-product. 

I thought about two ways:
 - Use clinker as the input of the add-on technology, then it produces more clinker. In this way, "efficiency improvement" is more or less modeled. But, it also means coal and electricity are reduced by the same percentage, which is not true for an add-on technology that saves only coal and even consumes more electricity. 
 - Use coal and electricity as input for the add-on technology, then it produces coal and electricity again, together with clinker... Technically, it could work if I design the parameters carefully. 

However, neither of the two looks good to me. I guess there could be better practice.

**Second, to model CCS technologies**, I thought about using CO2 as input of the add-on technology and it produces nothing. This sounds good to me. But I am not sure how this interacts with the emission accounting in MESSAGE. Suggestions are most welcomed!

Thanks for taking the time to answer my questions! If there are existing open-source that I can learn, you can direct me there briefly.

**Comment by glatterf42** [comment]
Hi @SongminYu and sorry for the late reply! Unfortunately, I'm not familiar with how similar things are modeled in our code, so I can't really suggest how to best model these add-on technologies. However, cement-related code was recently made open source via [message-ix-models/material](https://github.com/iiasa/message-ix-models/tree/main/message_ix_models/model/material), maybe you can find some information there. Or maybe @macflo8 could jump in here and provide some guidance :pray: 

For the second part, I imagine @ywpratama is the best person to come in and share experience. Even without having modeled CCS myself, I imagine you would have to have your CCS-add-on consume energy, too, otherwise, what's keeping the model from simply removing all CO2 with that without any cost?

**Comment by macflo8** [comment, marked as answer]
Hi @SongminYu,
if the technologies that you want to add are retrofittable in existing clinker kilns the `addon` feature of `MESSAGEix` seems like the best choice. If they require a completely new kiln, you might just model them as a regular `technology`.
In the case of modelling as an addon:
Your first options sounds good to me in principle. To avoid the same relative reduction of electricity and coal per unit of clinker produced, you can add an extra `input` of either electricity or coal to your `addon` `technology` to balance out the efficiency gains to the desired levels.

Regarding a CCS addon:
If you do not have CO2 as a `commodity` in your model, you can do it via the `emission_factor` parametrization. We have done this in the past by giving the CCS addon a negative value for `emission_factor`. This needs to be reflected in your reporting procedure of the model results then of course. So you would need to sum the emissions of these two technologies in the end to get the correct CO2 emissions from `clinker_making`.

**Comment by SongminYu** [comment]
Hi @glatterf42, hi @macflo8, thanks for the suggestions! I tried implementing the `clinker_making_ccs` addon technology. As you suggested, I parametrized its `emission_factor` with negative value and assumed `addon_conversion = 1` for it. It seems working perfectly. 

There is one thing I want to ask and confirm: I don't fully understand the `addon_conversion` parameter. For example, in the [tutorial](https://github.com/iiasa/message_ix/blob/main/tutorial/westeros/westeros_addon_technologies.ipynb), when adding `po_turbine` to `coal_ppl`, the `addon_conversion` is calculated as the inverse of the input of `po_turbine` --> $1/0.2 = 5$. 

My questions are:

- First, could you elaborate a bit more about the `addon_conversion` parameter? What does it mean? Why is it calculated in this way in the tutorial?
- Second, for the `addon technologies` of efficiency measures and CCS, is it right to assume the values as 1? I did this because I read the the following on the "[parameter definition](https://docs.messageix.org/en/latest/model/MESSAGE/parameter_def.html#input-output-mapping-costs-and-engineering-specifications)" page: _No default addon_conversion factor (conversion factor between add-on and parent technology activity) is set. This is to avoid default conversion factors of 1 being set for technologies with multiple modes, of which only a single mode should be linked to the add-on technology._ 

Besides, another question about modeling "efficiency measures" with addon technologies: can I parametrize their `input` with negative values? That would be a lot easier. 

Thanks!

**Reply by ywpratama** [reply]
Hi @SongminYu, I think for `addon_conversion`, this [documentation](https://docs.messageix.org/en/latest/model/MESSAGE/model_core.html#constraints-for-addon-technologies) here can describe what the parameter does. Basically, this parameter is to set the level of activity an `addon` technology will do corresponding to the parent's activity. 

For your second question, it's perhaps easier to just give an example:
For parent technology, each 1 Mt of cement production generates 10 Mt of CO2 (I just made up these numbers).
Now, you have a CCS addon which captures 90% of the CO2 emissions. Hence, the `addon_conversion` will be 9, so that each 1 Mt of cement production, your addon will capture 9 Mt of CO2. 

I hope this helps.

**Reply by SongminYu** [reply]
Hi @ywpratama, thanks for your answer! Following this example, if I already defined the `emission_factor` of my CCS addon technology to be $-9$, do I still use `addon_conversion = 9`? I thought in such a case I should define `addon_conversion = 1`.

As I understand, if I account `CO2` as a `commodity` in the model, which is the `input` of the `addon technology`. Then, your example defining `addon_conversion = 9` would work as expected. However, if I define the `emission_factor` of the addon as $-9$ (the CCS addon only has electricity as input), it will capture 9 Mt of CO2 with one unit of `ACT`. 

I chose to model the CO2 emission in this way (not as a `commodity`) and use the negative emission factor because I want to implement the CO2 bound with the `bound_emission` parameter.

**Reply by glatterf42** [reply]
Maybe one addition to understanding `addon_conversion`, if I understand the docs correctly: if your `addon` technology consumes electricity as an input, it could theoretically get that from all technologies of your model producing electricity. By designating it as an `addon` technology, it will only be able to use what is produced by its "parent technology". And `addon_conversion` is then used to formalize this and define how efficiently that can happen. 
In the model variable terms, you have $`\mathrm{ACT}_\mathrm{addon} = \mathrm{addon\_conversion} \cdot \mathrm{ACT}_\mathrm{parent}`$. So (using the tutorial example) when 1 unit of parent `ACT` enables 5 units of addon `ACT`, the conversion factor needs to be 5. In the tutorial case, this aligns with the inverse of the `addon` input, because this value was already used to formalize the `addon` technology's efficiency: output set to 1 divided by input should reflect the efficiency.

**Comment by ywpratama** [comment]
Hi @SongminYu , thanks for this question. I just wanted to add some more options here, in addition to what @glatterf42 and @macflo8 have provided. For partial efficiency improvements, e.g., reduce electricity input, you can perhaps add an `addon` technology that generates electricity "for free" for your parent technology (i.e., electricity as output) for your `addon` and becomes input for the parent. likewise, if it makes electricity consumption becomes more efficient but require more coal, then you can include coal as input. 

For your second question, for CCS technologies as addon, it will depends on how your model does the carbon accounting. In this context, I would second @macflo8 comment. generally, putting negative values for emission factor for the addon technology would work.
